workers/task-driver/src/tasks/pay_relayer_fee.rs

```diff
@@ -0,0 +1,211 @@
+    ProofGeneration(String),
+    /// An error interacting with the state
+    State(String),
+    /// An error updating validity proofs after the fees are settled
+    UpdateValidityProofs(String),
+}
+
+impl TaskError for PayRelayerFeeTaskError {
+    fn retryable(&self) -> bool {
+        match self {
+            PayRelayerFeeTaskError::Arbitrum(_)
+            | PayRelayerFeeTaskError::ProofGeneration(_)
+            | PayRelayerFeeTaskError::State(_)
+            | PayRelayerFeeTaskError::UpdateValidityProofs(_) => true,
+        }
+    }
+}
+
+impl Display for PayRelayerFeeTaskError {
+    fn fmt(&self, f: &mut Formatter<'_>) -> FmtResult {
+        write!(f, "{self:?}")
+    }
+}
+
+impl Error for PayRelayerFeeTaskError {}
+
+impl From<StateError> for PayRelayerFeeTaskError {
+    fn from(err: StateError) -> Self {
+        PayRelayerFeeTaskError::State(err.to_string())
+    }
+}
+
+// -------------------
+// | Task Definition |
+// -------------------
+
+/// Defines the pay fees task
+pub struct PayRelayerFeeTask {
+    /// The balance to pay fees for
+    pub mint: BigUint,
+    /// The wallet that this task pays fees for
+    pub old_wallet: Wallet,
+    /// The new wallet after fees have been paid
+    pub new_wallet: Wallet,
+    /// The proof of `VALID RELAYER FEE SETTLEMENT` used to pay the protocol fee
+    pub protocol_proof: Option<RelayerFeeSettlementBundle>,
+    /// The arbitrum client used for submitting transactions
+    pub arbitrum_client: ArbitrumClient,
+    /// A hand to the global state
+    pub state: State,
+    /// The work queue for the proof manager
+    pub proof_queue: ProofManagerQueue,
+    /// A sender to the network manager's queue
+    pub network_sender: NetworkManagerQueue,
+    /// The current state of the task
+    pub task_state: PayRelayerFeeTaskState,
+}
+
+#[async_trait]
+impl Task for PayRelayerFeeTask {
+    type State = PayRelayerFeeTaskState;
+    type Error = PayRelayerFeeTaskError;
+    type Descriptor = PayRelayerFeeTaskDescriptor;
+
+    async fn new(descriptor: Self::Descriptor, ctx: TaskContext) -> Result<Self, Self::Error> {
+        let old_wallet = ctx
+            .state
+            .get_wallet(&descriptor.wallet_id)?
+            .ok_or_else(|| PayRelayerFeeTaskError::State(WALLET_NOT_FOUND.to_string()))?;
+        let new_wallet = Self::get_new_wallet(&descriptor.balance_mint, &old_wallet)?;
+
+        Ok(Self {
+            mint: descriptor.balance_mint,
+            old_wallet,
+            new_wallet,
+            protocol_proof: None,
+            arbitrum_client: ctx.arbitrum_client,
+            state: ctx.state,
+            proof_queue: ctx.proof_queue,
+            network_sender: ctx.network_queue,
+            task_state: PayRelayerFeeTaskState::Pending,
+        })
+    }
+
+    #[allow(clippy::blocks_in_conditions)]
+    #[instrument(skip_all, err, fields(task = self.name(), state = %self.state()))]
+    async fn step(&mut self) -> Result<(), Self::Error> {
```

Comment:
calling out for when you fill in this stub: would be nice to tally up a "fees collected" metric here

Reply:
will add this to the implementation PR, thanks